input-profile.ts

```diff
@@ -0,0 +1,369 @@
+     * Returns url of input profile json file
+     */
+    url!: string;
+
+    /**
+     * A set of components to filter during component retrieval.
+     */
+    toFilter: Set<string> = new Set(['vr-mode-active-mode-switch']);
+
+    /**
+     * The default 3D controller model used when a custom model fails to load.
+     */
+    @property.object()
+    defaultController!: Object3D;
+
+    /**
+     * The index representing the handedness of the controller (0 for left, 1 for right).
+     */
+    @property.enum(hands, 0)
+    handednessIndex: number = 0;
+
+    /**
+     * The base path where XR input profiles are stored.
+     */
+    @property.string(
+        'https://cdn.jsdelivr.net/npm/@webxr-input-profiles/assets@latest/dist/profiles/'
+    )
+    path!: string;
+
+    /**
+     * An optional folder path for loading custom XR input profiles.
+     */
+    @property.string()
+    customProfileFolder!: string;
+
+    /**
+     * If true, the input profile will be mapped to the default controller, and no dynamic 3D model of controller will be loaded.
+     */
+    @property.bool(false)
+    mapToDefaultController!: boolean;
+
+    /**
+     * The object which has HandTracking component added to it.
+     */
+    @property.object()
+    trackedHand!: Object3D;
+
+    /**
+     * If true, adds a VR mode switch component to the loaded controller model.
+     */
+    @property.bool(true)
+    addVrModeSwitch!: boolean;
+    start() {
+        this._controllerModel = null;
+        this.toFilter = new Set(['vr-mode-active-mode-switch']);
+        this._defaultControllerComponents = this._getComponents(this.defaultController);
+        this._handedness = hands[this.handednessIndex];
+
+        this.engine.onXRSessionStart.add(() => {
+            this.engine.xr?.session.addEventListener(
+                'inputsourceschange',
+                this._onInputSourcesChange.bind(this)
+            );
+        });
+    }
+
+    onDeactivate() {
+        this.engine.xr?.session?.removeEventListener(
+            'inputsourceschange',
+            this._onInputSourcesChange.bind(this)
+        );
+    }
+
+    /**
+     * Sets newly loaded controllers for the HandTracking component to proper switching.
+     * @param {Object3D} controllerObject - The controller object.
+     * @hidden
+     */
+
+    private _setHandTrackingControllers(controllerObject: Object3D) {
+        const handtrackingComponent = this.trackedHand.getComponent(HandTracking);
+        if (!handtrackingComponent) return;
+        /** @todo: Remove any when HandTracking is typed. */
+        (handtrackingComponent as any).controllerToDeactivate = controllerObject;
+    }
+
+    /**
+     * Retrieves all components from the specified object and its children.
+     * @param {Object3D} obj - The object to retrieve components from.
+     * @return {Component[]} An array of components.
+     * @hidden
+     */
+    private _getComponents(obj: Object3D) {
+        if (obj == null) return;
```

Comment:
Typescript should tell you this is not possible, because the type is `Object3D` not `Object3D|null`

Reply:
here we are passing defaultContoller a editor assignable property as the obj , if its not assigned it could be null.
maybe should have done 
```js
private _getComponents(obj: Object3D | null)
```
but anyway i restructured it so that it has a default fallback w.r.t the editor template structure

```js
this.defaultController = this.defaultController || this.object.children[0];
```

but for custom project with different structure it could still become null 

Can we go with  

```js
	private _getComponents(obj: Object3D) {
		const components: Component[] = [];
		if (obj == null) return components;
```

in that case ?

Reply:
Yes absolutely.